### PR TITLE
feat: support loading types from ethPM Manifest

### PR DESCRIPTION
**Pull request by fubuloubu**
### What I did
Allow loading ABI types from ethPM manifest files

### How I did it
Hacking

### How to verify it
The whole imports system needs a lot of work

### Description for the changelog
Load interfaces from an ethPM Manifest file

### Cute Animal Picture

peakaboo

### CODE REVIEW

File: vyper/cli/vyper_compile.py [hunk 1]
```diff
@@ -217,10 +217,22 @@ def get_interface_codes(root_path: Path, contract_sources: ContractCodes) -> Dic
             with valid_path.open() as fh:
                 code = fh.read()
                 if valid_path.suffix == ".json":
-                    interfaces[file_path][interface_name] = {
-                        "type": "json",
-                        "code": json.loads(code.encode()),
-                    }
+                    contents = json.loads(code.encode())
+                    if "contractTypes" in contents:
```
Comment (charles-cooper) [inline vyper/cli/vyper_compile.py:221]: add a comment explaining this is handling ethPM v3 files

**Review by charles-cooper — CHANGES_REQUESTED** [review]

File: vyper/cli/vyper_compile.py (hunk 1 above)
Comment (fubuloubu) [inline vyper/cli/vyper_compile.py:221]:
```suggestion
                    # EthPM Manifest (EIP-2678)
                    if "contractTypes" in contents:
```

**Comment by charles-cooper** [comment]
also maybe add a test

File: vyper/cli/vyper_compile.py [hunk 2]
```diff
@@ -217,10 +217,22 @@ def get_interface_codes(root_path: Path, contract_sources: ContractCodes) -> Dic
...
+                        interfaces[file_path][interface_name] = {
+                            "type": "json",
+                            "code": contents["contractTypes"][interface_name]["abi"],
+                        }
+
+                    elif "abi" in contents or isinstance(contents, list):
```
Comment (fubuloubu) [inline vyper/cli/vyper_compile.py:227]:
```suggestion
                    # ABI JSON file (either `{"abi": List[ABI]}` or `List[ABI]`)
                    elif "abi" in contents or isinstance(contents, list):
```

File: vyper/cli/utils.py [hunk 3]
```diff
@@ -13,6 +13,12 @@ def get_interface_file_path(base_paths: Sequence, import_path: str) -> Path:
         suffix = next((i for i in (".vy", ".json") if file_path.with_suffix(i).exists()), None)
         if suffix:
             return file_path.with_suffix(suffix)
+
+        # Handle ethPM Manifests
```
Comment (fubuloubu) [inline vyper/cli/utils.py:17]:
```suggestion
        # Handle ethPM Manifests (`from path.to.Manifest import InterfaceName`)
```

**Comment by fubuloubu** [comment]
Tested locally by using Brownie (which uses JSON mode to compile) and downloaded copy of OpenZeppelin/openzeppelin-contracts converted to ethPM manifest format (using Ape)

**Comment by fubuloubu** [comment]
Like to note this code is really hacky and should be refactored so it finds the different types of interfaces faster:
```py
from location import Bar  # can either be `location/Bar.{vy|json}` or `location.json:Bar`
```

**Review by charles-cooper — APPROVED** [review]

**Pull request merged**